dask/array/routines.py

```diff
@@ -1190,3 +1191,118 @@ def insert(arr, obj, values, axis):
     interleaved = list(interleave([split_arr, split_values]))
     interleaved = [i for i in interleaved if i.nbytes]
     return concatenate(interleaved, axis=axis)
+
+
+def _einsum_kernel(*operands, **kwargs):
+    subscripts = kwargs.pop('subscripts')
+    ncontract_inds = kwargs.pop('ncontract_inds')
+    dtype = kwargs.pop('kernel_dtype')
+    chunk = np.einsum(subscripts, *operands, dtype=dtype, **kwargs)
+
+    # Avoid concatenate=True in atop by adding 1's
+    # for the contracted dimensions
+    return chunk.reshape(chunk.shape + (1,) * ncontract_inds)
+
+
+@wraps(np.einsum)
+def einsum(subscripts, *operands, **kwargs):
+    casting = kwargs.get('casting', 'safe')
+    dtype = kwargs.get('dtype')
+    optimize = kwargs.get('optimize', True)
+    order = kwargs.get('order', 'K')
+    einsum_dtype = dtype
+
+    if (not isinstance(subscripts, string_types) or
+            any(isinstance(o, string_types) for o in operands)):
+
+        raise ValueError("einsum(op0, sublist0, "
+                         "op1, sublist1, "
+                         "..., [sublistout]) "
+                         "call style is not "
+                         "currently supported.")
+
+    # Infer the output dtype from operands
+    if dtype is None:
+        dtype = np.result_type(*[o.dtype for o in operands])
+
+    if optimize is None:
+        optimize = False
+
+    can_optimize = LooseVersion(np.__version__) >= LooseVersion("1.12.0")
+
+    if can_optimize and optimize is not False:
+        optimize, path_info = np.einsum_path(subscripts, *operands,
+                                             optimize=optimize)
+
+        # We can obtain a more complete subscript specification
+        # from the path_info string, as it will fill in
+        # missing outputs and replace ellipses with concrete
+        # subscripts.
+        contraction_str = 'Complete contraction:'
+        start = path_info.find(contraction_str) + len(contraction_str)
+        end = path_info.find('\n')
+        subscripts = path_info[start:end].strip()
```

Comment:
Honestly, I'd rather require the more explicit syntax than parse this from a string, which seems quite fragile.

The other option is to copy `_parse_einsum_input()` from NumPy, which conveniently has the same license as dask (yay open source!):
https://github.com/numpy/numpy/blob/8323be1bc44c2811fc36f5b99c1a30ebcee8edbd/numpy/core/einsumfunc.py#L374

I think that function could work basically unchanged for dask, with `asanyarray` referring to the dask.array version (or maybe removed if unnecessary).

Reply:
I was loathe to let go of the full functionality so I've gone with the `_parse_einsum_input()` approach.